hip/base/hipsparse_bindings.hip.hpp

```diff
@@ -189,6 +189,131 @@ GKO_BIND_HIPSPARSE32_SPMV(ValueType, detail::not_implemented);
 #undef GKO_BIND_HIPSPARSE32_SPMV
 
 
+template <typename IndexType, typename ValueType>
+void spgemm_buffer_size(
+    hipsparseHandle_t handle, IndexType m, IndexType n, IndexType k,
+    const ValueType *alpha, const hipsparseMatDescr_t descrA, IndexType nnzA,
+    const IndexType *csrRowPtrA, const IndexType *csrColIndA,
+    const hipsparseMatDescr_t descrB, IndexType nnzB,
+    const IndexType *csrRowPtrB, const IndexType *csrColIndB,
+    const ValueType *beta, const hipsparseMatDescr_t descrD, IndexType nnzD,
+    const IndexType *csrRowPtrD, const IndexType *csrColIndD,
+    csrgemm2Info_t info, size_type &result) GKO_NOT_IMPLEMENTED;
+
+#define GKO_BIND_HIPSPARSE_SPGEMM_BUFFER_SIZE(ValueType, HipsparseName)        \
+    template <>                                                                \
+    inline void spgemm_buffer_size<int32, ValueType>(                          \
+        hipsparseHandle_t handle, int32 m, int32 n, int32 k,                   \
+        const ValueType *alpha, const hipsparseMatDescr_t descrA, int32 nnzA,  \
+        const int32 *csrRowPtrA, const int32 *csrColIndA,                      \
+        const hipsparseMatDescr_t descrB, int32 nnzB, const int32 *csrRowPtrB, \
+        const int32 *csrColIndB, const ValueType *beta,                        \
+        const hipsparseMatDescr_t descrD, int32 nnzD, const int32 *csrRowPtrD, \
+        const int32 *csrColIndD, csrgemm2Info_t info, size_type &result)       \
+    {                                                                          \
+        GKO_ASSERT_NO_HIPSPARSE_ERRORS(HipsparseName(                          \
+            handle, m, n, k, as_hiplibs_type(alpha), descrA, nnzA, csrRowPtrA, \
+            csrColIndA, descrB, nnzB, csrRowPtrB, csrColIndB,                  \
+            as_hiplibs_type(beta), descrD, nnzD, csrRowPtrD, csrColIndD, info, \
+            &result));                                                         \
+    }                                                                          \
+    static_assert(true,                                                        \
+                  "This assert is used to counter the false positive extra "   \
+                  "semi-colon warnings")
+
+GKO_BIND_HIPSPARSE_SPGEMM_BUFFER_SIZE(float, hipsparseScsrgemm2_bufferSizeExt);
+GKO_BIND_HIPSPARSE_SPGEMM_BUFFER_SIZE(double, hipsparseDcsrgemm2_bufferSizeExt);
+// TODO: uncomment when the shipped hipSparse gets updated
+// GKO_BIND_HIPSPARSE_SPGEMM_BUFFER_SIZE(std::complex<float>,
+//                                       hipsparseCcsrgemm2_bufferSizeExt);
+// GKO_BIND_HIPSPARSE_SPGEMM_BUFFER_SIZE(std::complex<double>,
+//                                       hipsparseZcsrgemm2_bufferSizeExt);*/
```

Comment:
AFAIK, this is supported in develop which has version 1.4.1. Should we add an `ifdef` for that?
https://github.com/ROCmSoftwarePlatform/hipSPARSE/blob/develop/CMakeLists.txt

Reply:
`grep Zcsrgemm /opt/rocm/hipsparse/include/hipsparse.h`:
In the `cuda-101-gnu8-llvm70` image the functions are declared, in `gnu7-llvm60` (AMD system) they are not present. We could definitely add an `ifdef`, I just wasn't sure when the new version will be packaged.

Reply:
I think the earliest support is in 1.4.0 according to 
https://github.com/ROCmSoftwarePlatform/hipSPARSE/blame/4a6a59259b4aeec84c4042b9ea0922a752c9e17a/CMakeLists.txt and https://github.com/ROCmSoftwarePlatform/hipSPARSE/pull/70

Reply:
The version that is currently packaged for debian still has version 1.3.2, so we'll definitely need this check.